README.md
# Simple message reader
> Convert `ArrayBuffer` or `Blob` to string (Browser only)

[![Travis](https://img.shields.io/travis/poppinss/simple-message-reader.svg?style=for-the-badge)](https://github.com/poppinss/simple-message-reader)

The websocket or XHR requests now can return data as a [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob) or [ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer). Using this package, you can conver them back to string.

```bash
npm i simple-message-reader
```

Pull from CDN

```html
<script src="https://unpkg.com/simple-message-reader"></script>
```

```js
import { isBlob, isArrayBuffer, parseBlob, parseArrayBuffer } from 'simple-message-reader'

// from CDN
const simpleMessageReader = window.simpleMessageReader
```

## API

## Constants

<dl>
<dt><a href="#acceptedTypesMap">acceptedTypesMap</a> : <code>Object</code></dt>
<dd><p>List of output formats accepted when parsing
blobs</p>
</dd>
</dl>

## Functions

<dl>
<dt><a href="#isBlob">isBlob(input)</a> ⇒ <code>Boolean</code></dt>
<dd><p>Returns a boolean telling if value is blob or
not</p>
</dd>
<dt><a href="#isArrayBuffer">isArrayBuffer(input)</a> ⇒ <code>Boolean</code></dt>
<dd><p>Returns a boolean telling if value is any of
the ArrayBuffer view or not</p>
</dd>
<dt><a href="#parseBlob">parseBlob(input, cb, [outputFormat])</a> ⇒ <code>void</code></dt>
<dd><p>Parses the blob to any of the accepted outputFormat types.</p>
<p>Callback will be called with error, if <code>outputFormat</code> is invalid
or <code>input</code> is not a blob.</p>
</dd>
<dt><a href="#parseArrayBuffer">parseArrayBuffer(input)</a> ⇒ <code>String</code></dt>
<dd><p>Parses the arrayBuffer to a string</p>
</dd>
</dl>

<a name="acceptedTypesMap"></a>

## acceptedTypesMap : <code>Object</code>
List of output formats accepted when parsing
blobs

**Kind**: global constant  
**Attribute**: acceptedTypesMap  
**Example**  
```js
['arrayBuffer', 'binaryString', 'dataURL', 'text']
```
<a name="isBlob"></a>

## isBlob(input) ⇒ <code>Boolean</code>
Returns a boolean telling if value is blob or
not

**Kind**: global function  

| Param | Type |
| --- | --- |
| input | <code>Blob</code> | 

**Example**  
```js
isBlob(new Blob())
```
<a name="isArrayBuffer"></a>

## isArrayBuffer(input) ⇒ <code>Boolean</code>
Returns a boolean telling if value is any of
the ArrayBuffer view or not

**Kind**: global function  

| Param | Type |
| --- | --- |
| input | <code>ArrayBufferView</code> | 

**Example**  
```js
isArrayBuffer(new Uint8Array())
```
<a name="parseBlob"></a>

## parseBlob(input, cb, [outputFormat]) ⇒ <code>void</code>
Parses the blob to any of the accepted outputFormat types.

Callback will be called with error, if `outputFormat` is invalid
or `input` is not a blob.

**Kind**: global function  

| Param | Type | Default |
| --- | --- | --- |
| input | <code>Blob</code> |  | 
| cb | <code>function</code> |  | 
| [outputFormat] | <code>String</code> | <code>&#x27;arrayBuffer&#x27;</code> | 

**Example**  
```js
parseBlob(new Blob(), function (error, output) {

}, 'arrayBuffer')
```
<a name="parseArrayBuffer"></a>

## parseArrayBuffer(input) ⇒ <code>String</code>
Parses the arrayBuffer to a string

**Kind**: global function  

| Param | Type |
| --- | --- |
| input | <code>ArrayBufferView</code> | 

**Example**  
```js
parseArrayBuffer(new Uint8Array([72, 101, 108, 108, 111, 32, 33])
```


CHANGELOG.md
<a name="1.0.4"></a>
## [1.0.4](https://github.com/poppinss/simple-message-reader/compare/v1.0.3...v1.0.4) (2018-03-11)



<a name="1.0.3"></a>
## [1.0.3](https://github.com/poppinss/simple-message-reader/compare/v1.0.2...v1.0.3) (2018-02-27)


### Bug Fixes

* **script:** fix prepublishOnly case ([9e55d39](https://github.com/poppinss/simple-message-reader/commit/9e55d39))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/poppinss/simple-message-reader/compare/v1.0.1...v1.0.2) (2018-02-27)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/poppinss/simple-message-reader/compare/v1.0.0...v1.0.1) (2018-02-27)


### Bug Fixes

* convert blob to arraybuffer ([e247f11](https://github.com/poppinss/simple-message-reader/commit/e247f11))


### Features

* change API ([ec6dfe2](https://github.com/poppinss/simple-message-reader/commit/ec6dfe2))



<a name="1.0.0"></a>
# 1.0.0 (2018-02-27)


### Features

* first working commit ([31a9a42](https://github.com/poppinss/simple-message-reader/commit/31a9a42))





test/array-buffer.spec.js
'use strict'

/**
 * websocket-message-reader
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
*/

import { isBlob, isArrayBuffer, parseBlob, parseArrayBuffer } from '../dist/smr.es.js'

group('Array buffer', () => {
  test('return true when input is blob', (assert) => {
    assert.isTrue(isBlob(new window.Blob()))
  })

  test('return true when input is Uint8Array', (assert) => {
    assert.isTrue(isArrayBuffer(new window.Uint8Array()))
  })

  test('read array buffer to string', (assert) => {
    const buff = new Uint8Array([72, 101, 108, 108, 111, 32, 33])
    const result = parseArrayBuffer(buff)
    assert.equal(result, 'Hello !')
  })

  test('read array buffer blob to Uint8Array', (assert, done) => {
    const buff = new window.Blob([new Uint8Array([72, 101, 108, 108, 111, 32, 33])])

    parseBlob(buff, (error, result) => {
      if (error) { return done(error) }
      assert.equal(parseArrayBuffer(result), 'Hello !')
      done()
    })
  })

  test('return error when outputFormat is not allowed', (assert, done) => {
    const buff = new window.Blob([new Uint8Array([72, 101, 108, 108, 111, 32, 33])])

    parseBlob(buff, (error, result) => {
      assert.equal(error.message, 'hello is not supported. Only arrayBuffer,binaryString,dataURL,text are allowed')
      done()
    }, 'hello')
  })

  test('return text from blob', (assert, done) => {
    const buff = new window.Blob([new Uint8Array([72, 101, 108, 108, 111, 32, 33])])

    parseBlob(buff, (error, result) => {
      if (error) { return done(error) }
      assert.equal(result, 'Hello !')
      done()
    }, 'text')
  })
})


bundle.js
!function(e,r){"object"==typeof exports&&"object"==typeof module?module.exports=r():"function"==typeof define&&define.amd?define([],r):"object"==typeof exports?exports.simpleMessageReader=r():e.simpleMessageReader=r()}(window,function(){return function(e){var r={};function n(t){if(r[t])return r[t].exports;var o=r[t]={i:t,l:!1,exports:{}};return e[t].call(o.exports,o,o.exports,n),o.l=!0,o.exports}return n.m=e,n.c=r,n.d=function(e,r,t){n.o(e,r)||Object.defineProperty(e,r,{configurable:!1,enumerable:!0,get:t})},n.r=function(e){Object.defineProperty(e,"__esModule",{value:!0})},n.n=function(e){var r=e&&e.__esModule?function(){return e.default}:function(){return e};return n.d(r,"a",r),r},n.o=function(e,r){return Object.prototype.hasOwnProperty.call(e,r)},n.p="",n(n.s=0)}([function(e,r,n){"use strict";Object.defineProperty(r,"__esModule",{value:!0}),r.isBlob=o,r.isArrayBuffer=function(e){return ArrayBuffer.isView(e)},r.parseBlob=function(e,r,n){var u=Object.keys(t);if(n=n||"arrayBuffer",-1===u.indexOf(n))return void r(new Error(n+" is not supported. Only "+u.join(",")+" are allowed"));if(!o(e))return void r(new Error("Input must be blob"));var i=new window.FileReader;i.onload=function(e){return r(null,e.target.result)},i.onerror=function(e){return r(e)},i.onabort=function(e){return r(e)},i[t[n]](e)},r.parseArrayBuffer=function(e){return String.fromCharCode.apply(null,new Uint8Array(e))};var t={arrayBuffer:"readAsArrayBuffer",binaryString:"readAsBinaryString",dataURL:"readAsDataURL",text:"readAsText"};function o(e){return e instanceof window.Blob==!0}}])});

karma.config.js
process.env.CHROME_BIN = require('puppeteer').executablePath()

// Karma configuration
// Generated on Sun Feb 25 2018 21:14:55 GMT+0530 (IST)

module.exports = function (config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['japa'],

    // list of files / patterns to load in the browser
    files: [
      {
        pattern: 'test/*.spec.js', watched: false
      }
    ],

    // list of files / patterns to exclude
    exclude: [
    ],

    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'test/*.spec.js': ['rollup']
    },

    rollupPreprocessor: {
      plugins: require('./rollup.plugins.js'),
      output: {
        format: 'iife',
        name: 'wspTest',
        sourcemap: 'inline'
      }
    },

    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],

    // web server port
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: false,

    browsers: ['Chrome_without_security'],

    customLaunchers: {
      Chrome_without_security: {
        base: 'ChromeHeadless',
        flags: ['--no-sandbox']
      }
    },

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: true,

    // Concurrency level
    // how many browser should be started simultaneous
    concurrency: Infinity
  })
}


index.js
'use strict'

/**
 * websocket-message-reader
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
*/

/**
 * List of output formats accepted when parsing
 * blobs
 *
 * @type {Object}
 *
 * @attribute acceptedTypesMap
 *
 * @example
 * ```js
 * ['arrayBuffer', 'binaryString', 'dataURL', 'text']
 * ```
 */
const acceptedTypesMap = {
  'arrayBuffer': 'readAsArrayBuffer',
  'binaryString': 'readAsBinaryString',
  'dataURL': 'readAsDataURL',
  'text': 'readAsText'
}

/**
 * Returns a boolean telling if value is blob or
 * not
 *
 * @method isBlob
 *
 * @param  {Blob}  input
 *
 * @return {Boolean}
 *
 * @example
 * ```js
 * isBlob(new Blob())
 * ```
 */
export function isBlob (input) {
  return input instanceof window.Blob === true
}

/**
 * Returns a boolean telling if value is any of
 * the ArrayBuffer view or not
 *
 * @method isArrayBuffer
 *
 * @param  {ArrayBufferView}      input
 *
 * @return {Boolean}
 *
 * @example
 * ```js
 * isArrayBuffer(new Uint8Array())
 * ```
 */
export function isArrayBuffer (input) {
  return ArrayBuffer.isView(input)
}

/**
 * Parses the blob to any of the accepted outputFormat types.
 *
 * Callback will be called with error, if `outputFormat` is invalid
 * or `input` is not a blob.
 *
 * @method parseBlob
 *
 * @param  {Blob}     input
 * @param  {Function} cb
 * @param  {String}   [outputFormat = 'arrayBuffer']
 *
 * @return {void}
 *
 * @example
 * ```js
 * parseBlob(new Blob(), function (error, output) {
 *
 * }, 'arrayBuffer')
 * ```
 */
export function parseBlob (input, cb, outputFormat) {
  const acceptedTypes = Object.keys(acceptedTypesMap)
  outputFormat = outputFormat || 'arrayBuffer'

  if (acceptedTypes.indexOf(outputFormat) === -1) {
    cb(new Error(`${outputFormat} is not supported. Only ${acceptedTypes.join(',')} are allowed`))
    return
  }

  if (!isBlob(input)) {
    cb(new Error('Input must be blob'))
    return
  }

  const reader = new window.FileReader()
  reader.onload = (event) => cb(null, event.target.result)
  reader.onerror = (event) => cb(event)
  reader.onabort = (event) => cb(event)
  reader[acceptedTypesMap[outputFormat]](input)
}

/**
 * Parses the arrayBuffer to a string
 *
 * @method parseArrayBuffer
 *
 * @param  {ArrayBufferView}         input
 *
 * @return {String}
 *
 * @example
 * ```js
 * parseArrayBuffer(new Uint8Array([72, 101, 108, 108, 111, 32, 33])
 * ```
 */
export function parseArrayBuffer (input) {
  return String.fromCharCode.apply(null, new Uint8Array(input))
}
